File: src/ThemeParkPlanner.Console/ThemePark.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using Challenge.Core;

namespace ThemeParkPlanner
{
    public class ThemePark : IReadOnlyThemePark
    {
        public int MaxHoursPerDay { get; private set; }

        public int MaxMinutesPerDay
        {
            get { return MaxHoursPerDay*Constants.MinutesPerHour; }
        }

        public List<Attraction> Attractions { get; private set; }

        private IReadOnlyCollection<Attraction> _readOnlyAttractions;

        public IReadOnlyCollection<Attraction> ReadOnlyAttractions
        {
            get
            {
                return _readOnlyAttractions ?? (_readOnlyAttractions
                    = new ReadOnlyCollection<Attraction>(Attractions));
            }
        }

        public ThemePark(IEnumerable<Attraction> attractions, int maxHoursPerDay)
        {
            Attractions = attractions.ToList();
            MaxHoursPerDay = maxHoursPerDay;
        }

        public List<Guest> Guests { get; private set; }

        public static ThemePark Read(TextReader reader)
        {
            var values = from x in reader.ReadLineAsync().Result.Split(' ')
                select x.ParseInteger();

            // ReSharper disable once PossibleMultipleEnumeration
            var attractionCount = values.ElementAt(0);

            // ReSharper disable once PossibleMultipleEnumeration
            var maxHoursPerDay = values.ElementAt(1);

            var attractions = Attraction.ReadAll(reader, attractionCount).ToList();

            var result = new ThemePark(attractions, maxHoursPerDay);

            var queryCount = reader.ReadLineAsync().Result.ParseInteger();

            result.Guests = Guest.ReadAll(reader, queryCount, result).ToList();

            return result;
        }
    }
}

File: src/ParkPlannerCpp/theme_park.cpp
#include "theme_park.h"
#include "constants.h"
#include "attraction.h"

#include <common\streamutil.h>

#include <string>

namespace challenges {

    namespace park {

        theme_park::theme_park(std::vector<attraction> const & attractions, int max_hours_per_day)
            : _max_hours_per_day(max_hours_per_day)
            , _attractions(attractions)
            , _guests()
        {
        }

        theme_park::~theme_park()
        {
        }

        int theme_park::max_hours_per_day() const {
            return _max_hours_per_day;
        }

        int theme_park::max_minutes_per_day() const {
            const int result = _max_hours_per_day*constants::minutes_per_hour;
            return result;
        }

        int theme_park::attraction_count() const {
            return static_cast<int>(_attractions.size());
        }

        attraction & theme_park::get_attraction(std::vector<attraction>::size_type index) {
            return *(_attractions.begin() + index);
        }

        std::shared_ptr<theme_park> theme_park::read(std::istream & is) {

            auto values = read_ints(is);

            auto attraction_count = values[0];
            auto max_hours_per_day = values[1];

            auto attractions = attraction::read_all(is, attraction_count);

            auto result = std::make_shared<theme_park>(attractions, max_hours_per_day);

            auto query_count = read_ints(is)[0];

            result->_guests = guest::read_all(is, query_count, result.get());

            return result;
        }
    }
}
